bin/cron/pardot_import

```diff
@@ -197,19 +188,189 @@ def main
+  # (Although they are NOT valid in the query string, which is where this data is going.)
+  prospects_payload_json_encoded = prospects_payload_json_encoded.gsub("+", "%2B")
+
+  "#{config[:pardot_url]}?prospects=#{prospects_payload_json_encoded}"
+end
+
+# Make an API request with Pardot authentication, including appending auth params and refreshing Pardot API key and
+# retrying if necessary
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_with_auth_retry(url)
+  # do the post to Pardot
+  post_request_with_auth(url)
+rescue InvalidApiKeyException
+  # If we fail with an invalid key, that probably means our API key (which is good for one hour) has expired. Try one
+  # time to request a new API key and try the post again. If that fails, that is a fatal error.
+  request_pardot_api_key
+  post_request_with_auth(url)
+end
+
+# Make an API request with Pardot authentication
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_request_with_auth(url)
+  request_pardot_api_key if $pardot_api_key.nil?
+  # add the API key and user key parameters to the URL
+  auth_url = append_auth_params_to_url(url)
+  doc = post_request(auth_url, {})
+  doc
+end
+
+# Make an API request
+# @param url [String] URL to post to - must already include Pardot auth params in query string
+# @param params [Hash] hash of POST params (may be empty hash)
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_request(url, params)
+  uri = URI(url)
+
+  # make the HTTP POST
+  response = Net::HTTP.post_form(uri, params)
+
+  # do common error handling for Pardot response
+  raise "Pardot request failed with HTTP " + response.code unless PARDOT_SUCCESS_HTTP_CODES.include?(response.code)
+
+  doc = Nokogiri::XML(response.body, &:noblanks)
+  raise "Pardot response did not return parsable XML" if doc.nil?
+
+  error_details = doc.xpath('/rsp/err').text
+  raise InvalidApiKeyException if error_details.include? "Invalid API key or user key"
+
+  status = doc.xpath('/rsp/@stat').text
+  raise "Pardot response did not include status" if status.nil?
+
+  doc
+end
+
+# Append standard Pardot auth parameters (per-session API key and fixed user key) to a
+# Pardot API request
+# @param url [String] URL to post to
+# @returns [String] URL with auth parameters appended
+def append_auth_params_to_url(url)
+  "#{url}&api_key=#{$pardot_api_key}&user_key=#{CDO.pardot_user_key}"
+end
+
+# Parse a Pardot XML response and raise an exception on the first error
+# if there is one.
+# @param doc [Nokogiri::XML] XML response from Pardot
+def raise_if_response_error(doc)
+  status = doc.xpath('/rsp/@stat').text
+  if status != "ok"
+    error_text = doc.xpath('/rsp/errors/*').first.try(:text)
+    error_text = "Unknown error" if error_text.empty?
+    log doc.to_s
+    log error_text
+    raise "Error in Pardot response: #{error_text}"
```

Comment:
Pardot's error response is not documented, but all `error_text` I've ever seen returned are fixed messages, so I think we are fine. No danger of it being a secret (none are passed). I haven't ever seen Pardot do something like render a specific email address into a response.